Framework/Nexus/inc/MantidNexus/NexusIOHelper.h

```diff
@@ -82,77 +86,110 @@ void callGetData(::NeXus::File &file, T &buf, bool close_file) {
 template <typename T>
 void callGetSlab(::NeXus::File &file, std::vector<T> &buf,
                  const std::vector<int64_t> &start,
-                 const std::vector<int64_t> &size, bool close_file) {
+                 const std::vector<int64_t> &size, const bool close_file) {
   file.getSlab(buf.data(), start, size);
   if (close_file)
     file.closeData();
 }
 
-/** Templated function to read any type of data and store it into another vector
- * type. If the two types are the same, the conversion is skipped.
+/** Templated function to read any type of vector and (potentially) convert it
+ * to another type. If the two types are the same, the conversion is skipped.
  */
-template <typename T, typename U>
-std::vector<T> readNexusAnyVector(::NeXus::File &file, size_t size,
-                                  bool close_file) {
-  if constexpr (sizeof(T) < sizeof(U)) {
+template <typename T, typename U, typename DowncastT>
+void doReadNexusAnyVector(std::vector<T> &out, ::NeXus::File &file,
+                          const size_t size, const bool close_file) {
+  if constexpr (sizeof(T) < sizeof(U) &&
+                !std::is_same_v<DowncastT, AllowDowncasting>) {
     if (close_file)
       file.closeData();
     throw std::runtime_error(
         "Downcasting is forbidden in NeXusIOHelper::readNexusAnyVector");
-  } else if constexpr (std::is_same<T, U>::value) {
-    std::vector<T> buf(size);
+  } else if constexpr (std::is_same_v<T, U>) {
     if (size > 0)
-      callGetData(file, buf, close_file);
-    return buf;
+      callGetData(file, out, close_file);
   } else {
-    std::vector<T> vec(size);
     if (size > 0) {
       std::vector<U> buf(size);
       callGetData(file, buf, close_file);
-      std::transform(buf.begin(), buf.end(), vec.begin(),
+      std::transform(buf.begin(), buf.end(), out.begin(),
```

Comment:
Not sure if we should really refer to this process as downcasting `static_cast`. Downcasting is most usually used to refer to casting a base type to a derived type (OO), which is not what's happening here. This is a conversion. Having said this I believe panadas does use downcasting to refer to the same process as is being applied here. Maybe `AllowNarrowing` would be better?